### DISCUSSION TITLE
MUI Autocomplete shared component validation

### DISCUSSION THREAD

**Discussion by Biz-e-Services** [react-hook-form/react-hook-form#3343] [Q&A]
I've managed to  implement the Autocomplete and have pretty much got it working apart from validation issue. I have a form with multiple Autocomplete's. I fetch API data and it populates the controls but when I tab through my tab controls using onSubmit function to switch tabs, it triggers validation on all other Autocomplete fields and presents with "Field is required" even though they have selected items in them. The only way I can progress through tabs is to add or remove a selected item as in touch the field in some way.

How do you validate against Autcomplete when Autocomplete is a shared component?

Using "react-hook-form": "^6.6.0" and  "joi": "^17.3.0",

My "Shared" AutoComplete.js file
```
import React from 'react';
import { Controller } from "react-hook-form";

// Material UI
import Autocomplete from '@material-ui/lab/Autocomplete';
import Checkbox from '@material-ui/core/Checkbox';
import CheckBoxIcon from '@material-ui/icons/CheckBox';
import CheckBoxOutlineBlankIcon from '@material-ui/icons/CheckBoxOutlineBlank';
import TextField from "@material-ui/core/TextField";

const icon = <CheckBoxOutlineBlankIcon fontSize="small" />;
const checkedIcon = <CheckBoxIcon fontSize="small" />;

const AutoComplete = ({ control, options, name, label }) => {

    const getOpObj = option => {
        if (!option._id) option = options.find(op => op._id === option);
        return option;
    };

    return (
        <Controller
            render={props => (
                <Autocomplete
                    {...props}
                    autoComplete
                    multiple
                    disableCloseOnSelect
                    options={options}
                    getOptionLabel={option => getOpObj(option).name}
                    getOptionSelected={(option, value) => {
                        if (!option) return option;
                        return option._id === getOpObj(value)._id;
                    }}
                    renderOption={(option, { selected }) => (
                        <React.Fragment>
                            <Checkbox
                                icon={icon}
                                checkedIcon={checkedIcon}
                                style={{ marginRight: 8 }}
                                checked={selected}
                            />
                            {option.name}
                        </React.Fragment>
                    )}
                    renderInput={params => (

                        <TextField
                            {...params}
                            label={label}
                            variant="standard"
                            placeholder="Options..."
                            error={!!error}
                            helperText={helperText}
                        />
                    )}
                    onChange={(_, data) => {
                        props.onChange(data);
                        // props.onBlur(data);
                    }}
                />
            )
            }
            name={name}
            control={control}
        />
    );
};

export default AutoComplete;
```

my Parent Form
```
...`
import AutoComplete from "../../components/AutoComplete";
...

-- Tab one
<AutoComplete
                      control={control}
                      options={safetyInvestigationOptions}
                      name="safetyInvestigation"
                      label="Type of Investigation"
                      error={!!errors.safetyInvestigation}
                      helperText={errors?.safetyInvestigation?.message}
                    />
... plus 5 others

-- Tab two
<AutoComplete
                      control={control}
                      options={environmentInvestigationOptions}
                      name="environmentInvestigation"
                      label="Type of Investigation"
                      error={!!errors.environmentInvestigation}
                      helperText={errors?.environmentInvestigation?.message}
                    />
                  </Grid>
... plus others
etc...
```

my Button onSubmit function
```
const onSubmit = requestDetail => {
    action(requestDetail);

    if (riskClassification !== 'Low') {

      if (valueTab === 0) setValueTab(1);
      if (valueTab === 1) setValueTab(2);
      if (valueTab === 2) setValueTab(3);


      if (valueTab === 3) history.push("./step4");
    } else {
      history.push("./step4")
    }

  };
```

**Comment by bluebill1049** [comment]
what do you mean by shared?

**Reply by Biz-e-Services** [reply]
Hi Bill
By shared I mean a common component as in I can use that component and reference it in the calling form multiple times. Like in your MuiAutoComplete example where you call it from Mui.js (just multiple times)

FormComponent.js
```'
...
<Grid item lg={12} xs={12}>
    <MuiAutoComplete 
        control={control}
        options={safetyInvestigationOptions}
        name="safetyInvestigation"
        label="Type of Investigation"
        error={!!errors.safetyInvestigation}
        helperText={errors?.safetyInvestigation?.message}
    />
</Grid>
<Grid item lg={12} xs={12}>
    <MuiAutoComplete 
        control={control}
        options={safetyInvestigationOptions}
        name="environmentInvestigation"
        label="Type of Investigation"
        error={!!errors.environmentInvestigation}
        helperText={errors?.environmentInvestigation?.message}
    />
</Grid>
<Grid item lg={12} xs={12}>
... etc over multiple tabs
</Grid>
...
```

**Reply by Biz-e-Services** [reply]
Just to add... I did a console.log on one of my field names on initial load it was set to undefined even though on the UI it was showing as being populated. I found that if I did a setValue (data from little-state-machine) in the useEffect it seemed to resolved the issue. Is this the correct resolve and is it optimal considering my code above?

**Reply by bluebill1049** [reply]
so you want to update both AutoComplete value? before going to the next page?

**Reply by Biz-e-Services** [reply]
Not necessarily. User may decide to save data as draft and then later finish, so these values could stay the same or may change. So typically the user may decide to leave values the same or change them. The will have to progress through each tab on wizard step 3 and then submit to step 4. Not sure if that answers your question?

**Reply by Biz-e-Services** [reply]
Sorry Bill I may have misunderstood your question... Form business logic as per my last comment. Form code is what I think you are asking.... yes I guess its to set/update the Autocomplete value.

**Reply by Biz-e-Services** [reply]
Hi Bill, still have issues with this. I've tries setting value= defaulValue= etc. but still getting...
```
A component is changing the uncontrolled value state of Autocomplete to be controlled.
```

I've tried implementing in at different places in code without resolve. Any help on this please?

**Reply by bluebill1049** [reply]
i think you are missing defaultValue, that's why that error message shows up

**Reply by Biz-e-Services** [reply]
as in 
```
defaultValue=""
```
I've tried using that on the calling <AutoComplete defaultValue="" and I've tried placing it in the shared component, but still get error.

**Reply by bluebill1049** [reply]
or null, can you share a simple CSB?

**Reply by Biz-e-Services** [reply]
Hi Bill. I created a code sandbox here https://codesandbox.io/s/holy-silence-cvhwu?file=/src/App.js
Not included all the code but I think enough to demonstrate the 
```
Material-UI: A component is changing the uncontrolled value state of Autocomplete to be controlled.
```
error.

Please let me know if the link does not work for you. I could not get a few things to work like the DevTool tag. Could not get state to work either, which is populated via API call on initial load of page 1 of wizard. How I resolved the "edit" (API call to existing document) is I used setValue() in useEffect... but for new record I get uncontrolled value error. (Hope this makes sense)

**Reply by Biz-e-Services** [reply]
When I setValues from my API call in the useEffect and comment out  defaultValues for just the wizard page where I have Autocomplete
```
useForm({
    //defaultValues: state.requestDetail,
   ...
  });
```
... it seems to work, but this seems to introduce a total re-render of whole page when navigating to other pages in wizard. Any idea as to why the setting of defaultValues causes the issue of "Material-UI: A component is changing the uncontrolled"?

Bill is there a place I can submit my project to, to have someone review my code to see if it's optimal and written correctly? Need to make sure I'm implementing my project in a wholesome way before continuing to write the rest of the app.

**Reply by Biz-e-Services** [reply]
Hi Bill
Back onto this again. I've updated my sandbox to explain and present my issue as above.  https://codesandbox.io/s/late-fire-4xem4?file=/src/Step3.js

If you go to tab 2 and press the next button you get errors on tab 1 which are populated via API call to database on initial load of page. I'm thinking it's to do with the state not persisting from tab 1 to tab 2 and vise versa... any ideas on how to fix?

**Reply by Biz-e-Services** [reply]
Hi Bill
I think I may have discovered the reason my validation is not working as expected. It seems my tabs are dropping off the DOM (Material UI) As JOI is validating the entire form it's not seeing the other tabs and throws validation errors. Is there a way to call resolver from within the form and not as below?

```
  const { control, errors, handleSubmit, setValue, watch, formState, register, clearErrors } = useForm({
    defaultValues: state.requestDetail,
    mode: "onChange",
    resolver: joiResolver(schema),  <--- this to be called on condition of what tab you are on and have the schema broken down into what appears on each tab
    // shouldUnregister: false
  });
```

so I'll have tab1_schem, tab2_schema, tab3_schems etc 

Please see https://codesandbox.io/s/late-fire-4xem4?file=/src/Step3.js for more code)

**Reply by bluebill1049** [reply]
even `shouldUnregister: false` doesn't work for you?

**Comment by Biz-e-Services** [comment]
No. When inspecting the DOM all tabs get removed except for the one you click on. Still presents same validation error on unfocused tabs.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by Biz-e-Services** [comment]
Actually Bill, I came across an article you have written in your FAQ section of the  rhf site dealing with tabs and models. Will follow that example. Ps. Have you noticed or been told that your rhf site is presenting white pages background...possibly css issues?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Reply by bluebill1049** [reply]
is this still an issue with the doc? (white) what's your dark mode setting? it shouldn't be a problem anymore if you just toggle to dark mode and it will persist.
